### PR TITLE
Allow dynamic configuration of maximal assertion message size.

### PR DESCRIPTION
**Pull request by orenbenkiki**
Resolve issue #127.

The new function is annotated as `@since 0.12.0` under the assumption that would be the next released version.

### CODE REVIEW

File: src/check_pack.c [hunk 1]
```diff
@@ -38,16 +38,33 @@
...
  * continuously reallocate) one big chunk for the whole message stream.
  * Problems were seen in the wild with up to 4 GB reallocations.
  */
 
+void check_set_max_msg_size(int max_msg_size) {
+    ck_max_msg_size = max_msg_size;
+}
+
+static int get_max_msg_size(void) {
+    if (ck_max_msg_size <= 0) {
```
Comment (brarcher) [inline src/check_pack.c]: To be consistent with the behavior of the other environment variables, the order of precedence should be:
1) environment variable
2) check_set_max_msg_size()
3) default value

See CK_DEFAULT_TIMEOUT as an example of this.

File: src/check_pack.c [hunk 2]
```diff
@@ -38,16 +38,33 @@
...
+        char *env = getenv("CK_MAX_MSG_SIZE");
+        if (env)
+            ck_max_msg_size = atoi(env);
+
+        if (ck_max_msg_size <= 0)
+            ck_max_msg_size = 4096;
```
Comment (brarcher) [inline src/check_pack.c]: Could 4096 still be a variable or ```#define``` above, so it is not a magic number?

**Comment by brarcher** [comment]
I think it would be valuable to document this new API call in the manual, maybe somewhere in [this](https://github.com/libcheck/check/blob/master/doc/check.texi#L988) section. Also mention that the environment variable can override the value.

There is an section which shows all the environment variables [here](https://github.com/libcheck/check/blob/master/doc/check.texi#L2250). Also mention the new environment variable there.

Finally, I think it would be valuable to have a new test for this, which shows that if the value is set too low Check does cause an error but if it is set high enough Check continues with the test. Many unit tests in Check run using Check itself, which verifies that certain errors occur when they are expected. The tests themselves are defined up to [this](https://github.com/libcheck/check/blob/master/tests/check_check_sub.c#L2706) point, they are added to the test suite (with whatever environment variables are used) [here](https://github.com/libcheck/check/blob/master/tests/check_check_sub.c#L2708), and the table which says what the output of the tests should be is [here](https://github.com/libcheck/check/blob/master/tests/check_check_master.c#L68). Try adding a test or so to verify the behavior of the new API and environment variable.

**Comment by orenbenkiki** [comment]
Thanks for the review pointers. I fixed the code to give precedence to the environment variable, and updated the TexInfo documentation.

I have difficulty with creating a test though. A too-long message invokes `eprintf`, which writes to `stderr`. It seems the expected-output mechanism doesn't capture such output... is there an example of a test that captures this sort of output I could use as a guide?

File: doc/check.texi [hunk 3]
```diff
@@ -993,6 +993,17 @@ repetitive code that is hard to read.  For your convenience Check
...
 conditions.
 
+@findex check_set_max_msg_size
+@vindex CK_MAX_MSG_SIZE
+The typical size of an assertion message is less than 80 bytes.
+However, some of the functions listed below can generate very large messages
+(up to 4GB allocations were seen in the wild).
+To prevent this, a limit is placed on the assertion message size.
+This limit is 4K bytes by default.
+It can be modified by setting the @code(CK_MAX_MSG_SIZE) environment variable,
```
Comment (brarcher) [inline doc/check.texi]: I think that you may need ```{}``` around ```@code```.

Comment (brarcher) [reply doc/check.texi]: That is:
```@code{CK_MAX_MSG_SIZE}```

**Comment by orenbenkiki** [comment]
Working on the tests; I see test/check_mem_leaks in the `tests/Makefile.am`, so I can add the new tests there. I don't see them listed anywhere else though, in particular in any of the `CMakeLists.txt` file. Is adding the new tests to `tests/Makefile.am` sufficient?

File: tests/check_env_max_msg_size.c [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+/**
+ * The purpose of this test is to be used by valgrind to check for
```
Comment (brarcher) [inline tests/check_env_max_msg_size.c]: May need to update this comment

File: tests/check_set_max_msg_size.c [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+/**
+ * The purpose of this test is to be used by valgrind to check for
```
Comment (brarcher) [inline tests/check_set_max_msg_size.c]: Same here, the comment needs to be updated

File: tests/check_set_max_msg_size.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+int main (void)
+{
+    int n;
+    SRunner *sr;
+
+    /*
+     * Run the test suite. This is intended to trigger the "Message is too long" error.
+     * Actual success/failure is determined by examining the output.
+     */
+    check_set_max_msg_size(32); // This will have an effect due to no environment variable.
```
Comment (brarcher) [inline tests/check_set_max_msg_size.c]: The two test programs could be collapsed by passing in a value over the command line, as I think that is the only difference, correct?

File: tests/test_env_max_msg_size.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env sh
```
Comment (brarcher) [inline tests/test_env_max_msg_size.sh]: The two test scripts could be collapsed, and simply run the unit test executable multiple times and check the result.

File: doc/check.texi [hunk 8]
```diff
@@ -993,6 +993,17 @@ repetitive code that is hard to read.  For your convenience Check
 provides a set of functions (actually macros) for testing often used
 conditions.
 
+@findex check_set_max_msg_size
```
Comment (brarcher) [inline doc/check.texi]: I think this could be moved to the end of the section, as it would not be expected to be called as frequently as the others. Or, do you disagree?

**Comment by brarcher** [comment]
Jenkins: ok to test
^ will start the remaining tests

**Comment by brarcher** [comment]
> Strange that it seems that cmake doesn't run the memory leak check though...?

The valgrind test is specific to Linux. The original intention of CMake was to support Windows targets, so it did not make much sense to add it. It turns out that some users on Linux do want to use CMake as well. 

Regardless, the unit tests are really to help protect the quality of the project. As long as the project still runs the tests on changes, I think being supported in one or the other build system should not be that important.

File: doc/check.texi (hunk 8 above)
Comment (orenbenkiki) [reply doc/check.texi]: I think that placing it at the end would make it "disappear" for someone who would say "ah, a list of all kind of assertions, I'll skip to the next section and come back here if I want to locate a specific assertion function". The other option is to make it an explicit section of its own but that seems an overkill for a single simple function.

**Comment by orenbenkiki** [comment]
Fixed everything (except for moving the documentation paragraph).

**Comment by orenbenkiki** [comment]
Ready to merge - or is there anything else?

File: src/check_pack.c [hunk 9]
```diff
@@ -38,16 +38,36 @@
...
- * is bigger than this (actually we check against CK_MAX_MSG_SIZE/2).
+ * is bigger than this.
  * The usual size for a message is less than 80 bytes.
  * All this is done instead of the previous approach to allocate (actually
  * continuously reallocate) one big chunk for the whole message stream.
  * Problems were seen in the wild with up to 4 GB reallocations.
  */
 
+void check_set_max_msg_size(int max_msg_size) {
+    if (ck_max_msg_size <= 0) {
```
Comment (brarcher) [inline src/check_pack.c]: Is it that once the value has been set it cannot be changed? The comment in the prototype says this is the intent: "if used... must be called once...". I am fine with the behavior, just curious as to the rationale.

File: src/check_pack.c [hunk 10]
```diff
@@ -38,16 +38,36 @@
...
  * side. When sending messages, we assure that no single message
- * is bigger than this (actually we check against CK_MAX_MSG_SIZE/2).
+ * is bigger than this.
  * The usual size for a message is less than 80 bytes.
  * All this is done instead of the previous approach to allocate (actually
  * continuously reallocate) one big chunk for the whole message stream.
  * Problems were seen in the wild with up to 4 GB reallocations.
  */
 
+void check_set_max_msg_size(int max_msg_size) {
```
Comment (brarcher) [inline src/check_pack.c]: Small thing, but can you match the style of the { in this file, namely:
```
void check_set_max_msg_size(int max_msg_size()
{
...
```

File: src/check_pack.c [hunk 11]
```diff
@@ -38,16 +38,36 @@
...
+            ck_max_msg_size = atoi(env);
+        if (ck_max_msg_size <= 0)
+            ck_max_msg_size = max_msg_size;
+        if (ck_max_msg_size <= 0)
+            ck_max_msg_size = DEFAULT_MAX_MSG_SIZE;
+    }
+}
+
+static int get_max_msg_size(void) {
+    if (ck_max_msg_size <= 0)
```
Comment (brarcher) [inline src/check_pack.c]: Why is this necessary? Should not the setter verify that the value is sane?

Is there a reason that ck_max_msg_size must be signed? Is it valid for it to be negative? Could it be unsigned?

File: src/check_pack.c (hunk 9 above)
Comment (orenbenkiki) [reply src/check_pack.c]: I was probably too focused on a specific use case. Fixed.

File: src/check_pack.c (hunk 10 above)
Comment (orenbenkiki) [reply src/check_pack.c]: Done.

File: src/check_pack.c (hunk 11 above)
Comment (orenbenkiki) [reply src/check_pack.c]: Changed type to `size_t`, which is probably the "right" type for a size.

**Comment by orenbenkiki** [comment]
Changed type of argument to `size_t`, allow modifying the size at any point, and fixed the style.

File: src/check_pack.c [hunk 12]
```diff
@@ -51,21 +51,22 @@ static int ck_max_msg_size = 0;
...
+    if (ck_max_msg_size == 0)
+        ck_max_msg_size = DEFAULT_MAX_MSG_SIZE;
 }
 
-static int get_max_msg_size(void) {
-    if (ck_max_msg_size <= 0)
-        check_set_max_msg_size(0);
+static size_t get_max_msg_size(void)
+{
+    if (ck_max_msg_size == 0)      // If check_set_max_msg_size was not called,
```
Comment (brarcher) [inline src/check_pack.c]: Should not the getter be making the decision as to if the environment variable exists or not, and let the setter only set the value passed? This way, if the environment variable changed since when the setter was called and the getter was called the correct value would be returned? e.g.:
```
void check_set_max_msg_size(size_t max_msg_size)
 {
     ck_max_msg_size = max_msg_size;
 }

void get_max_msg_size()
{
    size_t value = 0;

    char *env = getenv("CK_MAX_MSG_SIZE");
     if (env)
         value = (size_t)strtoul(env, NULL, 10); // Cast in case size_t != unsigned long.
     if(value == 0)
         value = ck_max_msg_size;
     if (value == 0)
         value = DEFAULT_MAX_MSG_SIZE;

   return value;
}
```

Comment (orenbenkiki) [reply src/check_pack.c]: Changing the environment variable at run time seems strange, but I guess it could happen. Also, I guess performance isn't an issue, as this is only used when an assertion fails...

**Comment by orenbenkiki** [comment]
Fixed.

**Comment by brarcher** [comment]
Thanks for the changeset! I've merge it in.

If you would like, you can submit a merge request adding an entry into the [AUTHORS](https://github.com/libcheck/check/blob/master/AUTHORS) file for yourself.
